db/db.go

```diff
@@ -0,0 +1,325 @@
+package db
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
+	"errors"
+
+	"fmt"
+
+	"github.com/boltdb/bolt"
+)
+
+var (
+	buildBucket   = []byte("build")
+	pendingBucket = []byte("pending")
+	statusBucket  = []byte("status")
+	outputBucket  = []byte("output")
+	shaBucket     = []byte("sha")
+	refBucket     = []byte("ref")
+)
+
+func validate(start, end int) error {
+	if start < 0 {
+		return fmt.Errorf("invalid argument start: %d", start)
+	}
+
+	if end < -1 {
+		return fmt.Errorf("invalid argument end: %d", end)
+	} else if end >= 0 && start > end {
+		return fmt.Errorf("invalid argument start: %d, end: %d", start, end)
+	}
+
+	return nil
+}
+
+// DB is the database api for ci system.
+type DB struct {
+	db *bolt.DB
+}
+
+// Open opens a database given path
+func Open(path string) (*DB, error) {
+	db, err := bolt.Open(path, 0600, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &DB{db: db}, nil
+}
+
+// Close the database.
+func (d *DB) Close() error {
+	return d.db.Close()
+}
+
+// itob returns an 8-byte big endian representation of v
+func itob(v uint64) []byte {
+	b := make([]byte, 8)
+	binary.BigEndian.PutUint64(b, uint64(v))
+	return b
+}
+
+// btoi returns uint64 of 8-byte big endian representation
+func btoi(b []byte) uint64 {
+	return binary.BigEndian.Uint64(b)
+}
+
+type tx struct {
+	*bolt.Tx
+}
+
+type txErr error
+
+func (t *tx) must(err error) {
+	if err != nil {
+		panic(txErr(err))
+	}
+}
+
+func handle(f func(t *tx) error) func(t *tx) error {
+	return func(t *tx) (err error) {
+		defer func() {
+			if r := recover(); r != nil {
+				fmt.Println("recover")
+				switch e := r.(type) {
+				case txErr:
```

Comment:
I ended up using
```
			if r := recover(); r != nil {
				switch e := r.(type) {
				case error:
					err = e
				default:
					err = fmt.Errorf("recovered panic: %v", e)
				}
			}
```
to make sure every panic is handled (according our discussion that ci server should not crash in any case).
Please let me know if you think we should panic if the type of `r` is not `error` :)